### DISCUSSION TITLE
Should pytest copy the input parameters to a parameterized fixture?

### DISCUSSION THREAD

**Discussion by lithomas1** [pytest-dev/pytest#12537] [Ideas] [answered]
In pytest, when using a parameterized fixture, it looks like pytest does not copy the input parameters to the fixture.
This means that the inputs to the parameterized fixture are essentially cached if ``request.param`` is directly returned from the function, making it possible to modify the fixture values (even if the scope is equal to function).

Example test case. 
```
import io
import pytest

@pytest.fixture(
    params=[io.StringIO()],
)
def bufs(request):
    return request.param


def test1(bufs):
    bufs.write("hi!")

    assert bufs.getvalue() == "hi!"

def test2(bufs):
    bufs.write("bye!")
    
    # Fails!!! since there is already stuff in bufs
    assert bufs.getvalue() == "bye!"
```

To prevent this, should pytest copy the input parameters to fixtures?
(The above test case will pass if you do ``copy.copy`` on ``request.param``)

**Comment by RonnyPfannschmidt** [comment, marked as answer]
No, Parameters ough@t to be immutable 

If it's something mutable, make it in a fixture

**Reply by lithomas1** [reply]
Thanks, is this documented somewhere?

I skimmed through https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#parametrizing-fixtures, and didn't see anything about mutability in there.
